src/Rings/MPolyMap/MPolyQuo.jl

```diff
@@ -0,0 +1,95 @@
+################################################################################
+#
+#  Domains of type MPolyQuo
+#
+################################################################################
+
+function _check_imgs_quo(R, S::NCRing, imgs)
+  for i in 1:n, j in 1:(i - 1)
+    @req imgs[i] * imgs[j] == imgs[j] * imgs[i] "Images $i and $j do not commute"
+  end
+  gensI = gens(modulus(R))
+  for g in gensI
+    @req iszero(evaluate(g, imgs)) "Morphism is not well-defined"
+  end
+  return nothing
+end
+
+# no check for commutative codomains
+function _check_imgs_quo(R, S::Ring, imgs) 
+  gensI = gens(modulus(R))
+  for g in gensI
+    @req iszero(evaluate(g, imgs)) "Morphism is not well-defined"
+  end
+end
+
+function hom(R::MPolyQuo, S::Ring, images::Vector; check::Bool = true)
+  n = ngens(R)
+  @req n == length(images) "Number of images must be $n"
+  # Now coerce into S or throw an error if not possible
+  imgs = _coerce(S, images)
+  if check
+    _check_imgs(S, imgs)
+    _check_homo(S, imgs)
+  end
+  return MPolyAnyMap(R, S, nothing, copy(imgs)) # copy because of #655
+end
+
+function hom(R::MPolyQuo, S::Ring, coeff_map, images::Vector; check::Bool = true)
+  n = ngens(R)
+  @req n == length(images) "Number of images must be $n"
+  # Now coerce into S or throw an error if not possible
+  imgs = _coerce(S, images)
+  if check
+    _check_imgs(S, imgs)
+    _check_homo(S, imgs)
+  end
+
+  return MPolyAnyMap(R, S, coeff_map, copy(imgs)) # copy because of #655
+end
+
+################################################################################
+#
+#  Evaluation
+#
+################################################################################
+
+function _evaluate_plain(F::MPolyAnyMap{<: MPolyQuo}, u)
+  return evaluate(lift(u), _images(F))
+end
+
+function _evaluate_general(F::MPolyAnyMap{<: MPolyQuo}, u)
+  if domain(F) === codomain(F)
+    return evaluate(map_coefficients(coefficient_map(F), lift(u),
+                                     parent = domain(F)), F.img_gens)
+  else
+    S = temp_ring(F)
+    if S !== nothing
+      evaluate(map_coefficients(coefficient_map(F), lift(u),
+                                parent = temp_ring), F.img_gens)
+    else
+      evaluate(map_coefficients(coefficient_map(F), lift(u)), F.img_gens)
+    end
+  end
```

Comment:
My personal preference is to be always explicit about return values and so always use `return`. Here, though, I'd also recommend to rewrite the code a bit to make it clearer what the common parts are:
```suggestion
  if domain(F) === codomain(F)
    coeff = map_coefficients(coefficient_map(F), lift(u); parent = domain(F))
  else
    S = temp_ring(F)
    if S !== nothing
      coeff = map_coefficients(coefficient_map(F), lift(u); parent = temp_ring)
    else
      coeff = map_coefficients(coefficient_map(F), lift(u))
    end
  end
  return evaluate(coeff, F.img_gens)
```

If my guess above is right and the second occurrence of `temp_ring` is meant to be `S`, then this could be rewritten further to:
```suggestion
  if domain(F) === codomain(F)
    S = domain(F)
  else
    S = temp_ring(F)
  end
  if S !== nothing
    coeff = map_coefficients(coefficient_map(F), lift(u); parent = S)
  else
    coeff = map_coefficients(coefficient_map(F), lift(u))
  end
  return evaluate(coeff, F.img_gens)
```

Reply:
Writing it in this way makes it possibly type unstable. I don't know what the individual `map_coefficients` will return, but I know that in each case evaluating will give something of the same type. The missing return statements were not intended :)